File: commands/song.js
const axios = require('axios');
const Discord = require('discord.js');

module.exports = {
  name: "song",
  args: true,
  alias: ["map"],
  description: "Gives details about the specified beatmap.",
  async execute(msg, args) {
    msg.channel.startTyping();

    if (args[0].toLowerCase() == "id") {
      msg.channel.send(await getSongByID(args[1])); // do what it says...
    } else if (args[0].toLowerCase() == "name") {
      msg.channel.send(await getSongByName(args)); // also do what it says...
    } else {
      msg.channel.send("No search type specified, defaulting to \"name\".");
      msg.channel.send(await getSongByName(args));
    }

    msg.channel.stopTyping();
  }
};

async function getSongByID(songID) {
  const config = {
    method: "get",
    url: `https://beatsaver.com/api/maps/detail/${songID}`,
    headers: {
      "Content-Type": "application/json",
      "User-Agent": "BeatSaberDiscordBot",
    },
  };

  let returnedEmbed;

  await axios(config)
    .then((res) => {
      const {
          metadata,
          stats,
          key,
          name,
          coverURL,
        } = res.data;
        const {
          difficulties,
          duration,
          levelAuthorName,
          bpm,
        } = metadata;
        const {
          downloads,
          downVotes,
          upVotes,
          rating,
        } = stats;
        const {
          easy,
          normal,
          hard,
          expert,
          expertPlus,
        } = difficulties;

        var difficultylist = "";
        for (var i in difficulties) { // Needs to be alphabetical I think, if anyone knows how feel free to contribute
          if (difficulties[i] == true && difficulties.hasOwnProperty(i)) {
            difficultylist += `- ${i.charAt(0).toUpperCase() + i.slice(1)}\n`;
          }
        }

        const mapEmbed = new Discord.MessageEmbed()
          .setColor("#f03030")
          .setTitle(`**Beatmap:** ${name}`)
          .setURL(`https://beatsaver.com/beatmap/${key}`)
          .setDescription(
            `📥 [One click install](http://spectrix.pythonanywhere.com/botsaber?key=${key}) with [ModAssistant](https://github.com/Assistant/ModAssistant)!
            👀 [Preview this map in your browser!](https://skystudioapps.com/bs-viewer/?id=${key}) `
          )
          .addFields(
            {
              name: "ℹ __Beatmap Info__",
              value: (
                `• **Level Author:** ${aiOrNo(levelAuthorName)}\n` +
                `• **Duration:** ${durationCheck(duration)}\n` +
                `• **Beatmap BPM:** ${Math.round(
                  (bpm + Number.EPSILON) * 100,
                ) / 100}`
              ),
            },
            {
              name: "📈 __Beatmap Stats__",
              value: (
                `• **Downloads:** ${downloads.toLocaleString()}\n` +
                `• **Upvotes:** ${upVotes.toLocaleString()}\n` +
                `• **Downvotes:** ${downVotes.toLocaleString()}\n` +
                `• **Rating:** ${Math.round(
                  (rating * 100 + Number.EPSILON) * 100,
                ) / 100}%`
              ),
            },
            {
              name: "📊 __Beatmap Difficulties__",
              value: (
                (difficultylist)
              ),
            },
          )
          .setThumbnail(`https://beatsaver.com${coverURL}`)
          .setFooter(
            `🔑 ${key} | Data from BeatSaver.`,
            "https://cdn.discordapp.com/attachments/478201257417244675/760182130352586802/unknown.png",
          );
	returnedEmbed = mapEmbed;
    });
  return returnedEmbed;
}

async function getSongByName(args) {
  if (args[0] == "name") {
    args.shift();
  }

  const config = {
    method: "get",
    url: `https://beatsaver.com/api/search/advanced/0?q=${
      encodeURI(args.join(" ").replace(/\\\//g, " ").replace(/\//g, " "))
    }`,
    headers: {
      "Content-Type": "application/json",
      "User-Agent": "BeatSaberDiscordBot",
    },
  };

  let returnedEmbed;

  await axios(config)
    .then(async (res) => {
      const { key } = res.data.docs[0];
      const embedSend = await getSongByID(key);
      
      returnedEmbed = embedSend;
    });

  return returnedEmbed;
}
function aiOrNo(input) {
  if (input == "Beat Sage") {
    return "Created by AI.";
  } else {
    return input;
  }
}

function pad_with_zeroes(number, length) {
    var my_string = '' + number;
    while (my_string.length < length) {
        my_string = '0' + my_string;
    }
    return my_string;
};

function durationCheck(input) {
  if (input == 0) {
    return "Not Specified.";
  } else {
    var minutes = Math.floor(input / 60);
    var seconds = input - minutes * 60;
    return `${minutes}:${pad_with_zeroes(seconds, 2)}`
  }
};

File: cogs/song.py
import json, discord, asyncio, json, math, aiohttp, datetime
from discord.ext import commands
from discord.ext.commands.cooldowns import BucketType
from disputils import BotEmbedPaginator
from discord import Embed

with open("config/thesacredtexts.json") as f:
    config = json.load(f)

headers = {
      "Content-Type": "application/json",
      "User-Agent": "BeatSaberDiscordBot",
    }

class Beatmap(commands.Cog):
    """ Commands that return info from beatmap(s) from ScoreSaber """
    def __init__(self, bot):
        self.bot = bot

    async def getSongInfo(self, result):
        metadata = result['metadata']
        stats = result['stats']
        embed = discord.Embed(title=f"**Beatmap:** {result['name']}", 
                                description=f":inbox_tray: [One click install](http://spectrix.pythonanywhere.com/botsaber?key={result['id']}) with [ModAssistant](https://github.com/Assistant/ModAssistant)!"+
                                            f"\n:eyes: [Preview this map in your browser!](https://skystudioapps.com/bs-viewer/?id={result['id']})",
                                color=0x0000ff, 
                                url=f"https://beatsaver.com/beatmap/{result['id']}")

        author = metadata['levelAuthorName'] + " (Automapping AI 🤖)" if metadata['levelAuthorName'] == "Beat Sage" else metadata['levelAuthorName']
        duration = str(datetime.timedelta(seconds=int(metadata['duration']))) if int(metadata['duration']) != 0 else "Not specified"
        duration = duration if duration[:1] != '0' else (duration[2:] if duration[2:][:1] != '0' else duration[3:])
        embed.add_field(name=":information_source: __Beatmap Info__",
                        value=f"• **Level Author:** {author}" +
                              f"\n• **Duration:** {duration}" +
                              f"\n• **Beatmap BPM:** {round(int(metadata['bpm']))}"
                              f"\n• **Ranked:** {result['ranked']}", inline=False)

        embed.add_field(name="📈 __Beatmap Stats__",
                        value="• **Downloads:** {:,d}".format(stats['downloads']) +
                        "\n• **Upvotes:** {:,d}".format(stats['upvotes']) +
                        "\n• **Downvotes:** {:,d}".format(stats['downvotes']) +
                        f"\n• **Rating:** {str(round(stats['score']*100))+'%' if stats['score']!=0 else 'Unrated'}", inline=False)

        # TODO: embed.add_field(name="📊 Beatmap Difficulties",

        if 1024 > len(result['description']) > 0:
            embed.add_field(name="📋 __Description__", value=f"```\n{result['description']}```", inline=False)
        
        
        embed.set_thumbnail(url=str(result['versions'][-1]['coverURL']))
        embed.set_footer(text=f"🔑 ID: {result['id']}", icon_url=config['styling']['logo'])
        return embed

    @commands.cooldown(1, 3, BucketType.user)
    @commands.group(aliases=['beatmap', 'map'])
    async def song(self, ctx):
        """Get information about a song from its id, or search for songs by name"""
        if ctx.invoked_subcommand is None:
            await ctx.send("**❌ Invalid input!** Ensure you provided full args. ```fix\nList of args:\n\n- name\n- id``` **For example:** `b!song name Midnight Lady`")


    @song.command()
    async def name(self, ctx, *, query):
        """Searches for songs with a specific name, returns a max of 10 results, to get song information"""
        async with ctx.typing():
            embeds = []
            async with aiohttp.ClientSession(headers=headers) as session:
                async with session.get(f"http://api.beatsaver.com/search/text/0?q={query}&sortOrder=Rating&automapper=true") as r:
                    result = await r.json()
            for i in range(len(result['docs']) if len(result['docs']) <= 10 else 10):
                embeds.append(await self.getSongInfo(result['docs'][i]))
        
        paginator = BotEmbedPaginator(ctx, embeds)

        try:
            await paginator.run()
        except IndexError:
            await ctx.send("**❌ No results found.** Think this is an error? Try again.")

    
    @song.command()
    async def id(self, ctx, id):
        """Lets you search for a song by its id to get song information"""
        async with ctx.typing():
            async with aiohttp.ClientSession(headers=headers) as session:
                async with session.get(f"http://api.beatsaver.com/maps/id/{id}") as r:
                    result = await r.json()
                    embed = await self.getSongInfo(result)
                    await ctx.send(embed=embed)     

    @commands.cooldown(1, 5, BucketType.user)
    @commands.command()
    async def latest(self, ctx):
        """Returns a few of the most recently published beatmaps on beatsaver"""
        embeds = []
        notify = await ctx.send("**<a:red_note:760170835729580065> Searching...** This takes time as I'm getting info from many songs...")

        async with ctx.typing():
            async with aiohttp.ClientSession(headers=headers) as session:
                async with session.get("http://api.beatsaver.com/maps/latest?automapper=true") as r:
                    result = await r.json()

            for i in result['docs']:
                embeds.append(await self.getSongInfo(i))
            
        
        paginator = BotEmbedPaginator(ctx, embeds)
        
        try:
            await notify.delete()
        except Exception:
            pass

        paginator = BotEmbedPaginator(ctx, embeds)

        
        await paginator.run() 

def setup(bot):
    bot.add_cog(Beatmap(bot))